<a href="https://colab.research.google.com/github/nmuthuk1/Python-Basics-HW/blob/master/Copy_of_spinal_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas dataframe exercises using the sprinter reaction time dataset.  
## Goal: Get familiar with basic dataframe operations

The code block below is a preamble that loads the libraries that we need and sets things up the way we like it.  You can copy and paste this preamble and put it in any notebook to have access to the same libraries.

In [0]:
# Tells the notebooks to view every graph "in line" automatically, rather than requiring a specific `show` command
%matplotlib inline 
# Loads the numpy library (for math stuff)
import numpy as np 
# Load the pandas library (what we use to create, explore, and analyze datasets)
import pandas as pd 
# Loads the matplotlib library (what we use to create graphs)
import matplotlib.pyplot as plt 
# Loads the seaborn library (what we use to create prettier graphs)
import seaborn as sns 
# Sets the default font size for graphs to 150% of its original value (for readability)
sns.set(font_scale=1.5) 

#### First we load the original data into a Pandas dataframe called `df`

In [0]:
url = "https://raw.githubusercontent.com/rgerkin/neuro-data-analysis-course/master/data/spinal/2004-2016-Short-Races2.csv"
df = pd.read_csv(url) # Loads the data from that URL into a dataframe
df.head() # Shows the first few rows of that dataframe

Unnamed: 0  POS  BIB          ATHLETE COUNTRY     MARK  RxnTime   Race  \
0           0  1.0  NaN       Usain BOLT  JAMJAM  10.20 Q    0.186  100 m   
1           1  2.0  NaN    Daniel BAILEY  ANTANT  10.24 Q    0.198  100 m   
2           2  3.0  NaN  Vicente DE LIMA  BRABRA  10.26 Q    0.168  100 m   
3           3  4.0  NaN   Henry VIZCAÍNO  CUBCUB  10.28 q    0.157  100 m   
4           4  5.0  NaN    Fabio CERUTTI  ITAITA    10.49    0.136  100 m   

   Round Gender  Year  
0  Heats      M  2008  
1  Heats      M  2008  
2  Heats      M  2008  
3  Heats      M  2008  
4  Heats      M  2008

#### Now complete the numbered tasks below.  Read the question, then use the code block to fill in your solution

1) A method is just a function that belongs to an object (like a dataframe).  The `.head()` method is a good way to preview the data that you loaded.  Sometimes people work with datasets with thousands or even millions of rows.  You usually don't need to see them all at once!  But sometimes you want to see more than 5 rows.  Try passing an argument to `.head()`, in order to see more rows.  For example, `.head(10)` to see 10 rows.  Or pick some other number.

In [0]:
# To visualize 500 rows of the data
df.head(500)

Unnamed: 0  POS     BIB                 ATHLETE COUNTRY     MARK  \
0             0  1.0     NaN              Usain BOLT  JAMJAM  10.20 Q   
1             1  2.0     NaN           Daniel BAILEY  ANTANT  10.24 Q   
2             2  3.0     NaN         Vicente DE LIMA  BRABRA  10.26 Q   
3             3  4.0     NaN          Henry VIZCAÍNO  CUBCUB  10.28 q   
4             4  5.0     NaN           Fabio CERUTTI  ITAITA    10.49   
5             5  6.0     NaN           Jurgen THEMEN  SURSUR    10.61   
6             6  7.0     NaN             Moses KAMUT  VANVAN    10.81   
7             7  8.0     NaN         Francis MANIORU  SOLSOL    11.09   
8             8  1.0     NaN            Asafa POWELL  JAMJAM  10.16 Q   
9             9  2.0     NaN             Kim COLLINS  SKNSKN  10.17 Q   
10           10  3.0     NaN         Craig PICKERING  GBRGBR  10.21 Q   
11           11  4.0     NaN           Daniel GRUESO  COLCOL  10.35 q   
12           12  5.0     NaN             Dariusz KUC  POLPOL  10.44 q   
13           13  6.0     NaN   Berenger Aymard BOSSE  CAFCAF    10.51   
14           14  7.0     NaN              Aisea TOHI  TGATGA    11.17   
15           15  8.0     NaN             Roman CRESS  MHLMHL    11.18   
16           16  1.0     NaN        Richard THOMPSON  TTOTTO  10.24 Q   
17           17  2.0     NaN       Martial MBANDJOCK  FRAFRA  10.26 Q   
18           18  3.0     NaN           Simone COLLIO  ITAITA  10.32 Q   
19           19  4.0     NaN             Aziz ZAKARI  GHAGHA  10.34 q   
20           20  5.0     NaN            Andrew HINDS  BARBAR  10.35 q   
21           21  6.0     NaN      Suryo Agung WIBOWO  INAINA    10.46   
22           22  7.0     NaN             Jared LEWIS  VINVIN       11   
23           23  8.0     NaN         Rabangaki NAWAI  KIRKIR    11.29   
24           24  1.0     NaN          Michael FRATER  JAMJAM  10.15 Q   
25           25  2.0     NaN           Pierre BROWNE  CANCAN  10.22 Q   
26           26  3.0     NaN            Darrel BROWN  TTOTTO  10.22 Q   
27           27  4.0     NaN        Nobuharu ASAHARA  JPNJPN  10.25 q   
28           28  5.0     NaN         Holder DA SILVA  GBSGBS    10.58   
29           29  6.0     NaN           Idrissa SANOU  BURBUR    10.63   
..          ...  ...     ...                     ...     ...      ...   
470         470  7.0  3106.0           Asenate MANOA  TUVTUV    13.48   
471         471  8.0  3363.0  Fatima Sulaiman DAHMAN  YEMYEM    13.95   
472         472  1.0  1330.0    Lorène Dorcas BAZOLO  CGOCGO  11.87 Q   
473         473  2.0  2036.0            Yee Pui FONG  HKGHKG  12.02 Q   
474         474  3.0  1410.0           Patricia TAEA  COKCOK    12.47   
475         475  4.0  3025.0          Maysa REJEPOVA  TKMTKM     12.8   
476         476  5.0  2944.0          Pauline KWALEA  SOLSOL     12.9   
477         477  6.0  2384.0     Laenly PHOUTTHAVONG  LAOLAO    13.15   
478         478  7.0  2640.0       Rubie Joy GABRIEL  PLWPLW    13.34   
479         479  NaN  2757.0   Noor Hussain AL-MALKI  QATQAT      DNF   
480         480  1.0  2642.0              Toea WISIL  PNGPNG  11.60 Q   
481         481  2.0  1798.0           Saruba COLLEY  GAMGAM  12.21 Q   
482         482  3.0  2942.0        Martina PRETELLI  SMRSMR    12.41   
483         483  4.0  1447.0           Lidiane LOPES  CPVCPV    12.72   
484         484  5.0  2410.0              Hala GEZAH  LBALBA    13.24   
485         485  6.0  2565.0           Pramila RIJAL  NEPNEP    13.33   
486         486  7.0  3343.0           Janice ALATOA  VANVAN     13.6   
487         487  8.0  1794.0         Mihter WENDOLIN  FSMFSM    13.67   
488         488  9.0  1002.0       Tahmina KOHISTANI  AFGAFG    14.42   
489         489  1.0  3055.0      Kelly-Ann BAPTISTE  TTOTTO  10.96 Q   
490         490  2.0  1791.0          Myriam SOUMARÉ  FRAFRA  11.07 Q   
491         491  3.0  1963.0           Verena SAILER  GERGER  11.12 Q   
492         492  4.0  2605.0        Ezinne OKPARAEBO  NORNOR  1

2) You can also look at the *last* few rows of the data, using the `.tail()` method. Try viewing the last 12 rows of the dataset.  

In [0]:
# To view the last 12 rows of the dataset
df.tail(12)

Unnamed: 0  POS  BIB               ATHLETE COUNTRY   MARK  RxnTime  \
2445        2445  3.0  NaN  Taiwo ALADEFA-DARDEN  NGRNGR  13.06      NaN   
2446        2446  4.0  NaN        Lidiya YURKOVA  BLRBLR   13.2      NaN   
2447        2447  5.0  NaN      Nadezhda BODROVA  UKRUKR  13.22      NaN   
2448        2448  6.0  NaN        Rachael ROGERS  FIJFIJ  14.07      NaN   
2449        2449  NaN  NaN      Elisabeta ANGHEL  ROUROU    DNF      NaN   
2450        2450  1.0  NaN           Gail DEVERS  USAUSA  12.73      NaN   
2451        2451  2.0  NaN  Gillian RUSSELL-LOVE  JAMJAM  12.85      NaN   
2452        2452  3.0  NaN        Yuliya GRAUDYN  RUSRUS  12.95      NaN   
2453        2453  4.0  NaN             Ime AKPAN  NGRNGR  13.11      NaN   
2454        2454  5.0  NaN       Olena KRASOVSKA  UKRUKR  13.23      NaN   
2455        2455  6.0  NaN     Véronique LINSTER  LUXLUX  13.47      NaN   
2456        2456  NaN  NaN           Carla TUZZI  ITAITA    DNF      NaN   

         Race  Round Gender  Year  
2445  100 m H  Heats      F  1996  
2446  100 m H  Heats      F  1996  
2447  100 m H  Heats      F  1996  
2448  100 m H  Heats      F  1996  
2449  100 m H  Heats      F  1996  
2450  100 m H  Heats      F  1996  
2451  100 m H  Heats      F  1996  
2452  100 m H  Heats      F  1996  
2453  100 m H  Heats      F  1996  
2454  100 m H  Heats      F  1996  
2455  100 m H  Heats      F  1996  
2456  100 m H  Heats      F  1996

3) How much data do you have?  You can find how many rows and columns you have with the `.shape` attribute.  Note the lack of parentheses here.  Attributes differ from methods in that they don't actually do anything (like print out information, or make a plot), but you can access their value to get information about your data.  The `.shape` attribute reports the number of rows and columns of the dataframe.  How many columns does the dataframe have?  

In [0]:
# Numbers of rows and coloumns in the dataset
df.shape

(2457, 11)

The data has 11 columns and 2457 rows in total.

4) Sometimes you only care about one column in your dataframe.  You can access it be name.  For example, `df['ATHLETE']` will access the "ATHLETE" column.  Access the column for reaction time. 

In [0]:
# Accessing groups/columns based on their title-Reaction time
df['RxnTime']
# df['Gender']

0       0.186
1       0.198
2       0.168
3       0.157
4       0.136
5       0.179
6       0.181
7       0.197
8       0.142
9       0.162
10      0.174
11      0.178
12      0.144
13      0.144
14      0.159
15      0.190
16      0.188
17      0.162
18      0.140
19      0.177
20      0.140
21      0.175
22      0.123
23      0.152
24      0.156
25      0.141
26      0.139
27      0.160
28      0.184
29      0.171
        ...  
2427      NaN
2428      NaN
2429      NaN
2430      NaN
2431      NaN
2432      NaN
2433      NaN
2434      NaN
2435      NaN
2436      NaN
2437      NaN
2438      NaN
2439      NaN
2440      NaN
2441      NaN
2442      NaN
2443      NaN
2444      NaN
2445      NaN
2446      NaN
2447      NaN
2448      NaN
2449      NaN
2450      NaN
2451      NaN
2452      NaN
2453      NaN
2454      NaN
2455      NaN
2456      NaN
Name: RxnTime, Length: 2457, dtype: float64

5) Notice that the above doesn't look like a dataframe anymore.  That's because you've only accessed one column, so the two-dimensional dataset (rows x columns) is now only one-dimensional. One-dimensional data in Pandas is convered into a so-called "Series".  Many of the same methods that belong to dataframes also belong to a series, e.g. `.head()` or `.mean()`.  Find the median reaction time as you did in the earlier homework assignment, but operating directly on the series containing the reaction times.  

In [0]:
# Median for the data series Reaction time
df['RxnTime'].median()

0.165

6) Sometimes you want to keep a dataframe, but just with fewer columns.  For example, only the name and reaction time.  You do this with an extra set of brackets, i.e. `df[['col1','col2']]` if you wanted to access hypothetical columns named 'col1' and 'col2'.  Try generating a dataframe that has only the name and reaction time columns from the original dataset.  

In [0]:
# To retain only name and reaction time from original dataset
df[['ATHLETE','RxnTime']]

ATHLETE  RxnTime
0                      Usain BOLT    0.186
1                   Daniel BAILEY    0.198
2                 Vicente DE LIMA    0.168
3                  Henry VIZCAÍNO    0.157
4                   Fabio CERUTTI    0.136
5                   Jurgen THEMEN    0.179
6                     Moses KAMUT    0.181
7                 Francis MANIORU    0.197
8                    Asafa POWELL    0.142
9                     Kim COLLINS    0.162
10                Craig PICKERING    0.174
11                  Daniel GRUESO    0.178
12                    Dariusz KUC    0.144
13          Berenger Aymard BOSSE    0.144
14                     Aisea TOHI    0.159
15                    Roman CRESS    0.190
16               Richard THOMPSON    0.188
17              Martial MBANDJOCK    0.162
18                  Simone COLLIO    0.140
19                    Aziz ZAKARI    0.177
20                   Andrew HINDS    0.140
21             Suryo Agung WIBOWO    0.175
22                    Jared LEWIS    0.123
23                Rabangaki NAWAI    0.152
24                 Michael FRATER    0.156
25                  Pierre BROWNE    0.141
26                   Darrel BROWN    0.139
27               Nobuharu ASAHARA    0.160
28                Holder DA SILVA    0.184
29                  Idrissa SANOU    0.171
...                           ...      ...
2427               Lesley TASHLIN      NaN
2428             Ludmila ENGQUIST      NaN
2429                 Angela ATEDE      NaN
2430        María José MARDOMINGO      NaN
2431  Monique ÉWANJÉ-ÉPÉE TOURRET      NaN
2432           Lena SOLLI-REIMANN      NaN
2433           Natalya GRIGORYEVA      NaN
2434               Sonia PAQUETTE      NaN
2435                Julie BAUMANN      NaN
2436                 Angela THORP      NaN
2437          Lynda TOLBERT-GOODE      NaN
2438             Kristin PATZWAHL      NaN
2439         Tatyana RESHETNIKOVA      NaN
2440                Cécile CINÉLU      NaN
2441                Bich Huong VU      NaN
2442               Aminata CAMARA      NaN
2443              Brigita BUKOVEC      NaN
2444           Dionne ROSE-HENLEY      NaN
2445         Taiwo ALADEFA-DARDEN      NaN
2446               Lidiya YURKOVA      NaN
2447             Nadezhda BODROVA      NaN
2448               Rachael ROGERS      NaN
2449             Elisabeta ANGHEL      NaN
2450                  Gail DEVERS      NaN
2451         Gillian RUSSELL-LOVE      NaN
2452               Yuliya GRAUDYN      NaN
2453                    Ime AKPAN      NaN
2454              Olena KRASOVSKA      NaN
2455            Véronique LINSTER      NaN
2456                  Carla TUZZI      NaN

[2457 rows x 2 columns]

7) Columns have names ('ATHLETE', 'COUNTRY', etc.), but so do rows.  The row names are part of the so-called "index" of the dataframe (or series), and allow you to access data more intuitively than just by row number.  In this case though, the names in the index are the row numbers.  You can extract row 144 with `df.loc[144]`.  Note the use of the `.loc` attribute, and the brackets instead of parentheses.  Use this extract the row containing Usain Bolt's 2008 preliminary heats data (hint, it is at the beginning of the dataframe).

In [0]:
# Extract the row with Usain Bolt's record
df.loc[0]

Unnamed: 0             0
POS                    1
BIB                  NaN
ATHLETE       Usain BOLT
COUNTRY           JAMJAM
MARK             10.20 Q
RxnTime            0.186
Race               100 m
Round              Heats
Gender                 M
Year                2008
Name: 0, dtype: object

8) You can access specific cells of the dataframe by row and column using `.loc`, e.g. `df.loc[7,'Country']`.  You can also edit data the same way.  Maybe Usain Bolt's preliminary heats time was faster than reported, say 9.5 seconds?  Edit the dataframe with `.loc` to make that change.  Be sure to use the `=` sign or you won't actually make the change!

In [0]:
# Edit Usain Bolt's record data
df.loc[0,'MARK'] = 9.5
df.loc[0,'MARK']

9.5

9) I don't like the way this data is organized.  Accessing it by arbitrary row numbers is not intuitive at all.  Let's organize the data by athlete name and year instead, using the `set_index()` method.  Inside the parentheses, put a list containing the column name indicating athlete name and the column name indicating the year.  Don't forget quotes and brackets!  

In [0]:
# Edit data organization based on column names
df.set_index(['ATHLETE','Year'])

Unnamed: 0  POS  BIB COUNTRY     MARK  \
ATHLETE                     Year                                          
Usain BOLT                  2008           0  1.0  NaN  JAMJAM      9.5   
Daniel BAILEY               2008           1  2.0  NaN  ANTANT  10.24 Q   
Vicente DE LIMA             2008           2  3.0  NaN  BRABRA  10.26 Q   
Henry VIZCAÍNO              2008           3  4.0  NaN  CUBCUB  10.28 q   
Fabio CERUTTI               2008           4  5.0  NaN  ITAITA    10.49   
Jurgen THEMEN               2008           5  6.0  NaN  SURSUR    10.61   
Moses KAMUT                 2008           6  7.0  NaN  VANVAN    10.81   
Francis MANIORU             2008           7  8.0  NaN  SOLSOL    11.09   
Asafa POWELL                2008           8  1.0  NaN  JAMJAM  10.16 Q   
Kim COLLINS                 2008           9  2.0  NaN  SKNSKN  10.17 Q   
Craig PICKERING             2008          10  3.0  NaN  GBRGBR  10.21 Q   
Daniel GRUESO               2008          11  4.0  NaN  COLCOL  10.35 q   
Dariusz KUC                 2008          12  5.0  NaN  POLPOL  10.44 q   
Berenger Aymard BOSSE       2008          13  6.0  NaN  CAFCAF    10.51   
Aisea TOHI                  2008          14  7.0  NaN  TGATGA    11.17   
Roman CRESS                 2008          15  8.0  NaN  MHLMHL    11.18   
Richard THOMPSON            2008          16  1.0  NaN  TTOTTO  10.24 Q   
Martial MBANDJOCK           2008          17  2.0  NaN  FRAFRA  10.26 Q   
Simone COLLIO               2008          18  3.0  NaN  ITAITA  10.32 Q   
Aziz ZAKARI                 2008          19  4.0  NaN  GHAGHA  10.34 q   
Andrew HINDS                2008          20  5.0  NaN  BARBAR  10.35 q   
Suryo Agung WIBOWO          2008          21  6.0  NaN  INAINA    10.46   
Jared LEWIS                 2008          22  7.0  NaN  VINVIN       11   
Rabangaki NAWAI             2008          23  8.0  NaN  KIRKIR    11.29   
Michael FRATER              2008          24  1.0  NaN  JAMJAM  10.15 Q   
Pierre BROWNE               2008          25  2.0  NaN  CANCAN  10.22 Q   
Darrel BROWN                2008          26  3.0  NaN  TTOTTO  10.22 Q   
Nobuharu ASAHARA            2008          27  4.0  NaN  JPNJPN  10.25 q   
Holder DA SILVA             2008          28  5.0  NaN  GBSGBS    10.58   
Idrissa SANOU               2008          29  6.0  NaN  BURBUR    10.63   
...                                      ...  ...  ...     ...      ...   
Lesley TASHLIN              1996        2427  7.0  NaN  CANCAN    13.61   
Ludmila ENGQUIST            1996        2428  1.0  NaN  SWESWE    12.66   
Angela ATEDE                1996        2429  2.0  NaN  NGRNGR    12.88   
María José MARDOMINGO       1996        2430  3.0  NaN  ESPESP    12.91   
Monique ÉWANJÉ-ÉPÉE TOURRET 1996        2431  4.0  NaN  FRAFRA    13.12   
Lena SOLLI-REIMANN          1996        2432  5.0  NaN  NORNOR    13.13   
Natalya GRIGORYEVA          1996        2433  6.0  NaN  UKRUKR    13.16   
Sonia PAQUETTE              1996        2434  7.0  NaN  CANCAN    13.29   
Julie BAUMANN               1996        2435  1.0  NaN  SUISUI    12.86   
Angela THORP                1996        2436  2.0  NaN  GBRGBR    12.93   
Lynda TOLBERT-GOODE         1996        2437  3.0  NaN  USAUSA    12.97   
Kristin PATZWAHL            1996        2438  4.0  NaN  GERGER    12.98   
Tatyana RESHETNIKOVA        1996        2439  5.0  NaN  RUSRUS    13.01   
Cécile CINÉLU               1996        2440  6.0  NaN  FRAFRA    13.05   
Bich Huong VU               1996        2441  7.0  NaN  VIEVIE    13.85   
Aminata CAMARA              1996        2442  8.0  NaN  MLIMLI    14.94   
Brigita BUKOVEC             1996        2443  1.0  NaN  SLOSLO    12.72   
Dionne ROSE-HENLEY          1996        2444  2.0  NaN  JAMJAM    12.81   
Taiwo ALADEFA-DARDEN        1996        2445  3.0  NaN  NGRNGR    13.06   
Lidiya YURKOVA              1996        2446  4.0  NaN  BLRBLR     13.2   
Nadezhda BODROVA            1996        2447  5.0  NaN  UKR

10) That's OK, but it doesn't look much nicer.  Maybe if it the index was sorted?  If you've ever done a hierarchical sort in Excel, this will be similar.  Try the same command as above, but append `.sort_index()`  to it

In [0]:
# Sort instead of Set in a particular order (Append)
df.set_index(['ATHLETE','Year']).sort_index()

Unnamed: 0  POS     BIB COUNTRY        MARK  \
ATHLETE                   Year                                                
Aaron BROWN               2016        1147  3.0  2192.0  CANCAN       10.24   
Abdul GHAFOOR             1996        2257  8.0     NaN  AFGAFG        12.2   
Abdul RASHID              2008         378  8.0     NaN  PAKPAK       14.52   
Abdul Wahab ZAHIRI        2016        1068  7.0  2000.0  AFGAFG       11.56   
Abdullah AL-SOOLI         2008          44  5.0     NaN  OMAOMA       10.53   
Abdullah Abkar MOHAMMED   2016        1106  5.0  2717.0  KSAKSA       10.26   
Abiodun OYEPITAN          2004        1316  2.0     NaN  GBRGBR     11.23 Q   
                          2004        1324  3.0     NaN  GBRGBR     11.28 Q   
                          2004        1358  5.0     NaN  GBRGBR       11.18   
                          2012         509  5.0  1869.0  GBRGBR     11.22 q   
                          2012         551  7.0  1869.0  GBRGBR       11.36   
Abraham KEPSIN            2000        1772  8.0     NaN  VANVAN       11.12   
Adam GEMILI               2012         639  2.0  1809.0  GBRGBR     10.11 Q   
                          2012         679  2.0  1809.0  GBRGBR       10.06   
Adrian DURANT             2004        1417  6.0     NaN  ISVISV       10.52   
Adrian GRIFFITH           2016        1091  8.0  2067.0  BAHBAH       10.53   
Adrian WOODLEY            2000        2037  4.0     NaN  CANCAN     13.71 Q   
                          2000        2064  6.0     NaN  CANCAN       14.04   
Adrien DEGHELT            2012         778  3.0  1136.0  BELBEL     13.52 Q   
                          2012         832  4.0  1136.0  BELBEL       13.42   
Afa ISMAIL                2012         460  5.0  2494.0  MDVMDV       12.52   
Affoué Amandine ALLOU     2004        1278  4.0     NaN  CIVCIV       11.46   
                          2008         174  5.0     NaN  CIVCIV       11.75   
Agné EGGERTH              2000        1894  6.0     NaN  LTULTU       11.87   
                          2004        1302  4.0     NaN  LTULTU       11.44   
Ahmad HAZER               2012         810  9.0  2421.0  LBNLBN       14.82   
                          2016        1208  7.0  2731.0  LIBLIB        15.5   
Ahmed DOUHOU              1996        2317  5.0     NaN  CIVCIV       10.53   
Aisea TOHI                2008          14  7.0     NaN  TGATGA       11.17   
Aissata TOURÉ             2012         462  7.0  2021.0  GUIGUI       13.25   
...                                    ...  ...     ...     ...         ...   
Yuniel HERNÁNDEZ          2004        1683  NaN     NaN  CUBCUB         DNS   
Yuriy ARISTOV             1996        2364  8.0     NaN  UZBUZB       15.04   
Yvette LEWIS              2016         933  8.0  1081.0  PANPAN       13.35   
Yvonne KANAZAWA           1996        2425  5.0     NaN  JPNJPN        13.3   
                          2000        2110  5.0     NaN  JPNJPN     13.13 q   
                          2000        2150  6.0     NaN  JPNJPN     13.11 q   
                          2000        2176  8.0     NaN  JPNJPN       13.16   
Zaidatul Husniah ZULKIFLI 2016         960  3.0   999.0  MASMAS     12.12 q   
                          2016        1029  8.0   999.0  MASMAS       12.62   
Zhanna BLOCK              1996        2192  1.0     NaN  UKRUKR        11.2   
                          2000        1873  1.0     NaN  UKRUKR     11.27 Q   
                          2000        1956  1.0     NaN  UKRUKR     11.08 Q   
                          2000        1974  3.0     NaN  UKRUKR     11.32 Q   
                          2000        1991  4.0     NaN  UKRUKR        11.2   
                          2004        1306  NaN     NaN  UKRUKR          DQ   
                          2004        1337  NaN     NaN  UKRUKR          DQ   
                          2004        1369  NaN     NaN  UKRUKR          DQ   
Zhenye XIE                2016        1102  1.0  2234.0  CHNCHN     10.08 Q   
   

11) Display the dataframe again by entering `df` on an empty line (e.g. in the code cell below, or create a new code cell).  Does it reflect the changes you made above?  If not, it's because `.set_index()`, `.sort_index()` (and most other methods) only return modified copies of your data; they don't overwrite it.  If you want to work with that modified copy, set the original dataframe equal to it, i.e. `df = ...` where `...` would contain the command you used above.  

In [0]:
# Index usage does not over-write
#df
# To set the sorted data as the original
df = df.set_index(['ATHLETE','Year']).sort_index()
df

Unnamed: 0  POS     BIB COUNTRY        MARK  \
ATHLETE                   Year                                                
Aaron BROWN               2016        1147  3.0  2192.0  CANCAN       10.24   
Abdul GHAFOOR             1996        2257  8.0     NaN  AFGAFG        12.2   
Abdul RASHID              2008         378  8.0     NaN  PAKPAK       14.52   
Abdul Wahab ZAHIRI        2016        1068  7.0  2000.0  AFGAFG       11.56   
Abdullah AL-SOOLI         2008          44  5.0     NaN  OMAOMA       10.53   
Abdullah Abkar MOHAMMED   2016        1106  5.0  2717.0  KSAKSA       10.26   
Abiodun OYEPITAN          2004        1316  2.0     NaN  GBRGBR     11.23 Q   
                          2004        1324  3.0     NaN  GBRGBR     11.28 Q   
                          2004        1358  5.0     NaN  GBRGBR       11.18   
                          2012         509  5.0  1869.0  GBRGBR     11.22 q   
                          2012         551  7.0  1869.0  GBRGBR       11.36   
Abraham KEPSIN            2000        1772  8.0     NaN  VANVAN       11.12   
Adam GEMILI               2012         639  2.0  1809.0  GBRGBR     10.11 Q   
                          2012         679  2.0  1809.0  GBRGBR       10.06   
Adrian DURANT             2004        1417  6.0     NaN  ISVISV       10.52   
Adrian GRIFFITH           2016        1091  8.0  2067.0  BAHBAH       10.53   
Adrian WOODLEY            2000        2037  4.0     NaN  CANCAN     13.71 Q   
                          2000        2064  6.0     NaN  CANCAN       14.04   
Adrien DEGHELT            2012         778  3.0  1136.0  BELBEL     13.52 Q   
                          2012         832  4.0  1136.0  BELBEL       13.42   
Afa ISMAIL                2012         460  5.0  2494.0  MDVMDV       12.52   
Affoué Amandine ALLOU     2004        1278  4.0     NaN  CIVCIV       11.46   
                          2008         174  5.0     NaN  CIVCIV       11.75   
Agné EGGERTH              2000        1894  6.0     NaN  LTULTU       11.87   
                          2004        1302  4.0     NaN  LTULTU       11.44   
Ahmad HAZER               2012         810  9.0  2421.0  LBNLBN       14.82   
                          2016        1208  7.0  2731.0  LIBLIB        15.5   
Ahmed DOUHOU              1996        2317  5.0     NaN  CIVCIV       10.53   
Aisea TOHI                2008          14  7.0     NaN  TGATGA       11.17   
Aissata TOURÉ             2012         462  7.0  2021.0  GUIGUI       13.25   
...                                    ...  ...     ...     ...         ...   
Yuniel HERNÁNDEZ          2004        1683  NaN     NaN  CUBCUB         DNS   
Yuriy ARISTOV             1996        2364  8.0     NaN  UZBUZB       15.04   
Yvette LEWIS              2016         933  8.0  1081.0  PANPAN       13.35   
Yvonne KANAZAWA           1996        2425  5.0     NaN  JPNJPN        13.3   
                          2000        2110  5.0     NaN  JPNJPN     13.13 q   
                          2000        2150  6.0     NaN  JPNJPN     13.11 q   
                          2000        2176  8.0     NaN  JPNJPN       13.16   
Zaidatul Husniah ZULKIFLI 2016         960  3.0   999.0  MASMAS     12.12 q   
                          2016        1029  8.0   999.0  MASMAS       12.62   
Zhanna BLOCK              1996        2192  1.0     NaN  UKRUKR        11.2   
                          2000        1873  1.0     NaN  UKRUKR     11.27 Q   
                          2000        1956  1.0     NaN  UKRUKR     11.08 Q   
                          2000        1974  3.0     NaN  UKRUKR     11.32 Q   
                          2000        1991  4.0     NaN  UKRUKR        11.2   
                          2004        1306  NaN     NaN  UKRUKR          DQ   
                          2004        1337  NaN     NaN  UKRUKR          DQ   
                          2004        1369  NaN     NaN  UKRUKR          DQ   
Zhenye XIE                2016        1102  1.0  2234.0  CHNCHN     10.08 Q   
   

Now that `df` has been changed, some of the commands in the earlier questions won't work the same.  So you won't be able to access e.g. `df.loc[7]` anymore, because there is no longer a row named 7!  This is why you should always execute notebooks from top to bottom.  Each notebook cell can change the data in a way that a cell above it doesn't know about.  

12) But now we have an intuitive index for the data frame.  Instead of accessing something like `df.loc[7]` we can access by name and year (using the same `.loc` attribute).  First just try accessing Usain Bolt's data by name (note that his last name is all capital letters in this dataset).  Note that it will contain results from several races across several years.  

In [0]:
# Accessing new sorted data with names
df.loc['Usain BOLT']

Unnamed: 0  POS     BIB COUNTRY     MARK  RxnTime   Race    Round Gender
Year                                                                          
2008           0  1.0     NaN  JAMJAM      9.5    0.186  100 m    Heats      M
2008         104  1.0     NaN  JAMJAM   9.92 Q    0.165  100 m  Quarter      M
2008         120  1.0     NaN  JAMJAM   9.85 Q    0.161  100 m     Semi      M
2008         136  1.0     NaN  JAMJAM     9.69    0.165  100 m    Final      M
2012         630  1.0  2170.0  JAMJAM  10.09 Q    0.178  100 m    Heats      M
2012         670  1.0  2170.0  JAMJAM   9.87 Q    0.180  100 m     Semi      M
2012         686  1.0  2170.0  JAMJAM     9.63    0.165  100 m    Final      M
2016        1136  1.0  2612.0  JAMJAM  10.07 Q    0.156  100 m    Heats      M
2016        1162  1.0  2612.0  JAMJAM   9.86 Q    0.143  100 m     Semi      M
2016        1178  1.0  2612.0  JAMJAM     9.81    0.155  100 m    Final      M

13) You can further restrict the data to a single year by using the `.loc` attribute again ("chaining" operations together), and now including a specific year.  Access Usain Bolt's data for 2008.  Hint: it will look just like the answer above, but with an extra `.loc` appended on.  

In [0]:
# Append .loc with year filter also
df.loc['Usain BOLT'].loc[2016]

Unnamed: 0  POS     BIB COUNTRY     MARK  RxnTime   Race  Round Gender
Year                                                                        
2016        1136  1.0  2612.0  JAMJAM  10.07 Q    0.156  100 m  Heats      M
2016        1162  1.0  2612.0  JAMJAM   9.86 Q    0.143  100 m   Semi      M
2016        1178  1.0  2612.0  JAMJAM     9.81    0.155  100 m  Final      M

14) How fast does Usain Bolt react to the starting gun?  Find his minimum Olympic sprint reaction time by acessing his data using `.loc`, accessing reaction time by column name, and then computing the minimum.  Hint: the *maximum* is obtained with the `.max()` method.  

In [0]:
# Minimum Reaction time for gun by Usain BOLT
df.loc[['Usain BOLT','RxnTime']].min()

Unnamed: 0         0
POS                1
BIB             2170
COUNTRY       JAMJAM
RxnTime        0.143
Race           100 m
Round          Final
Gender             M
dtype: object

15) Usain Bolt was the fastest 100m sprinter of all time, but he didn't try his hardest until the finals of each meet (he didn't need to).  Let's extract only his data from the finals.  This takes only one step, but we'll do it in two to make the idea clear.  First, let's generate a series that tells us whether it is True or False that a given race was a final.  Just paste `df['Round'] == 'Final'` in the cell below.  You will see a series containig some `True` values and some `False` values (mostly the latter), reflecting rows containing data from finals vs. other rounds.  Note the ellipsis halfway down, which represents all the data (hundreds of rows) it isn't showing you to save space.  

In [0]:
# To find out the finals
df1=df['Round'] =='Final'
df1.loc['Usain BOLT']

Year
2008    False
2008    False
2008    False
2008     True
2012    False
2012    False
2012     True
2016    False
2016    False
2016     True
Name: Round, dtype: bool

16) The expression above can be used as a "selector", to select data from the dataframe which matches the desired condition (i.e. only 200 m races).  Place it inside brackets (i.e. inside `df[]`) to access only the 200 m races from the original data.  

In [0]:
# Only 100m races from original
df ['Race'] == '100 m'


ATHLETE                    Year
Aaron BROWN                2016     True
Abdul GHAFOOR              1996     True
Abdul RASHID               2008    False
Abdul Wahab ZAHIRI         2016     True
Abdullah AL-SOOLI          2008     True
Abdullah Abkar MOHAMMED    2016     True
Abiodun OYEPITAN           2004     True
                           2004     True
                           2004     True
                           2012     True
                           2012     True
Abraham KEPSIN             2000     True
Adam GEMILI                2012     True
                           2012     True
Adrian DURANT              2004     True
Adrian GRIFFITH            2016     True
Adrian WOODLEY             2000    False
                           2000    False
Adrien DEGHELT             2012    False
                           2012    False
Afa ISMAIL                 2012     True
Affoué Amandine ALLOU      2004     True
                           2008     True
Agné EGGERTH             

17) Now use the above, plus chaining with `.loc[]` and `max()` to find Usain Bolt's *worst* reaction time in an Olympics final.   

In [0]:
# Usain BOLT's worst reaction time in the Olympics
df2 =df ['Round'] == 'Final'
df[df2].loc['Usain BOLT'].max()['RxnTime']

0.165

18) Let's clean up this data a little bit more.  See all the `NaN` values above?  Those correspond to missing data (NaN stands for "Not a Number").  We can get rid of rows that have missing data with the `.dropna()` method.  But some missing data is irrelevant to our analysis.  Who cares if the bib number is missing?  Select only those columns we care about (let's say reaction time, round, and gender) using the approach from question 6, and then use `dropna()` on that subset of the columns.  Chain that with `.shape` to see how many rows the remaining data has.  It should be less than the number of rows we started with (because we removed some rows), but still well more than 0!

In [0]:
# Get rid of NAN
df.dropna()['BIB']
df.shape

(2457, 9)

19) Do Americans sprinters react faster than Canadian sprinters?  No one would ever fund such a study, but you can answer that question right here for free!  Use the `.groupby()` method to aggregate sprinters by country `.groupby('COUNTRY')`, select the reaction time column, and chain it with `.mean()` to compare mean reaction time across countries.  You can further optionally append `[['USAUSA','CANCAN']]` to focus on only those two countries.  Why are all the country names doubled?  I think it is because the person who compiled this dataset likes to say, "U-S-A!  U-S-A!" and just repeated that for every country.  

In [0]:
# Comparing means of Rxntime for USA and Canada
df.groupby('COUNTRY')['RxnTime'].mean()[['USAUSA','CANCAN']]

COUNTRY
USAUSA    0.170185
CANCAN    0.159889
Name: RxnTime, dtype: float64

In [0]:
# Be sure to run this code cell!
from scipy.stats import ttest_ind

20) So *do* Americans react faster (or slower) than Canadians? In the previous question we only computed the means.  What about the variability?  Let's assume for the moment that a t-test is a good way to compare the two groups.  Extract the two groups using the data using the selector approach from question 16, extract the reaction time from each, drop the missing values, and assigning one to the variable `usa` and the other to the variable `can`.  Don't use `groupby()` here.  Having loaded a t-test function (`ttest_ind` in the cell above), call that function using your selected groups (`ttest_ind(usa,can)`).  Note that you can have multiple command in a code cell, as long each one is on its own line.  

In [0]:
# Fill in your answer to 20 here
USA = df ['COUNTRY'] == 'USAUSA'
CAN = df ['COUNTRY'] == 'CANCAN'
usa = df[USA]['RxnTime'].notnull()
can = df[CAN]['RxnTime'].notnull()
#usa1 = df[USA]['RxnTime'].dropna()
#can1 = df[CAN]['RxnTime'].dropna()
(ttest_ind(usa,can))


Ttest_indResult(statistic=0.6446917853701551, pvalue=0.5196804516037232)

21) Oh my, are we going to win a Nobel Prize for this discovery?  Or have we failed to control for something?  Maybe one country has more male vs female athletes, or more sprinters vs hurdlers.  Repeat the above, restricting the data to men (21a), then to women (21b), then to only 100 m sprinters (21c).  You can have multiple `==` expressions in your selector, but you should use parentheses to keep them separated e.g. `(df['COUNTRY']=='USA') && (df['Gender']=='F')`.  

In [0]:
# Restricting the test to men 
maleusa = (df['COUNTRY']=='USAUSA') & (df['Gender']=='M') # Use & to combine or logical &
malecan = (df['COUNTRY']=='CANCAN') & (df['Gender']=='M')
m_usa = df[maleusa]['RxnTime'].notnull()
m_can = df[malecan]['RxnTime'].notnull()
#m_usa1 = df[maleusa]['RxnTime'].dropna()
#m_can1 = df[malecan]['RxnTime'].dropna()
(ttest_ind(m_usa,m_can))

Ttest_indResult(statistic=1.1015188257161757, pvalue=0.27261656160210646)

In [0]:
# Restricting the test to women
femaleusa = (df['COUNTRY']=='USAUSA') & (df['Gender']=='F') # Use & to combine or logical &
femalecan = (df['COUNTRY']=='CANCAN') & (df['Gender']=='F')
f_usa = df[femaleusa]['RxnTime'].notnull()
f_can = df[femalecan]['RxnTime'].notnull()
#f_usa1 = df[femaleusa]['RxnTime'].dropna()
#f_can1 = df[femalecan]['RxnTime'].dropna()
(ttest_ind(f_usa,f_can))

Ttest_indResult(statistic=-0.14000961837077358, pvalue=0.8888707602578478)

In [0]:
# Restricting the test to 100 m sprinters
husa = (df['COUNTRY']=='USAUSA') & (df['Race']=='100 m') # Use & to combine or logical and (only 1 & symbol)
hcan = (df['COUNTRY']=='CANCAN') & (df['Race']=='100 m')
h_usa = df[husa]['RxnTime'].notnull()
h_can = df[hcan]['RxnTime'].notnull()
#h_usa1 = df[husa]['RxnTime'].dropna()
#h_can1 = df[hcan]['RxnTime'].dropna()
(ttest_ind(h_usa,h_can))

Ttest_indResult(statistic=0.8921243904320407, pvalue=0.37393748684496064)

### Congratulations, you are now familiar with the basic of programmatic operations on structured data!